### ISSUE TITLE
[BUG] Backup data in S3 is not deleted after the VM Backup is deleted from Harvester

### ISSUE THREAD

**Issue by johnliu55tw** [harvester/harvester#1333]
**Describe the bug**
Data still exist in S3 even if the VM Backup is deleted from Harvester. I further deleted the VM but data still exists in S3.

**To Reproduce**
1. Create a new VM
2. Log into the VM and create a 50M file
3. Create a Backup by "Take Backup"
4. Go to "Backups" page and wait for the backup I created to Ready
5. Click "YAML" to find the volumeName: `pvc-47d4726c-966c-4b69-bb89-527b9d6c07e4` and verify the folder with the name exists on S3
7. Delete the backup. The folder still exists on S3
8. Delete the VM. The folder still exists on S3
![image](https://user-images.githubusercontent.com/5047724/135403602-3b35b588-c92d-4d8f-a803-bb3963770859.png)

**Expected behavior**
The backup data of the volume should be removed from S3.

**Support bundle**
[supportbundle_cc9ca2f2-df88-46fc-b711-652252d0d1d1_2021-09-30T07-05-31Z.zip](https://github.com/harvester/harvester/files/7257378/supportbundle_cc9ca2f2-df88-46fc-b711-652252d0d1d1_2021-09-30T07-05-31Z.zip)

**Environment:**
 - Harvester ISO version: v0.3.0-rc1
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): KVM

**Additional context**
I checked Longhorn UI and even though the volume of the VM is gone, the backup still exists:

**No `pvc-47d4726c-966c-4b69-bb89-527b9d6c07e4` volume:**
![image](https://user-images.githubusercontent.com/5047724/135404175-94dda0c0-d87a-4c95-9548-5bd612d416fc.png)

**But its backup still exists**
![image](https://user-images.githubusercontent.com/5047724/135404123-8e25fdac-8c3c-4131-8969-4b0f111736dd.png)

If I deleted the backup from Longhorn UI, backup data disappeared from S3 entirely.

**Comment by lanfon72** [comment]
Reproducible on test case *Delete multiple backups* which is **Backup and Restore**'s No.8.

**Comment by khushboo-rancher** [comment]
Tried with v1.0.0-rc2, the blocks get deleted however the .cfg file doesn't get deleted.

**Comment by w13915984028** [comment]
@FrankYang0529 : Hello Frank, could you take a quick look and leave a short description about current specification of this behavior, will all related files are removed from backup target?  thanks.

**Comment by FrankYang0529** [comment]
> 

Yeah, I think most of the files will be removed if `VolumeSnapshotContent.DeletionPolicy` is `Delete`. LH Backup CR will also be removed if the policy is `Delete`. And then [longhorn-manager/controller/backup_controller.go](https://github.com/longhorn/longhorn-manager/blob/878cfb868c568396d6ebfa4ce096c5d95d9b31e3/controller/backup_controller.go#L237-L271) will be triggered. In [line 254](https://github.com/longhorn/longhorn-manager/blob/878cfb868c568396d6ebfa4ce096c5d95d9b31e3/controller/backup_controller.go#L254), it will trigger [`backup rm`](https://github.com/longhorn/backupstore/blob/master/cmd/remove.go) command line to delete files.

**Comment by w13915984028** [comment]
Test in Harvester v1.0.0, with NFS backup target:

1. Create a VM with 2 volumes, and after ready, take a "backup", wait ready,

in the NFS server, the `"pvc-7f1f8921-a2e9-4b96-bafa-12e080237bc2"` and related data is there
```
rancher@ubuntuvmday0:~$ sudo tree /nfs_server/nfs_share/backupstore/volumes
/nfs_server/nfs_share/backupstore/volumes
├── 17
│   └── 91
│       └── pvc-0a7cd9e3-b847-481b-a369-9f22686dee8f
│           ├── backups
│           │   └── backup_backup-29a244649f4b45e5.cfg
│           └── volume.cfg
├── 2d
│   └── d2
│       └── pvc-4bc23108-d2cb-44b8-8617-6d11248b2593
│           └── volume.cfg
├── 36
│   └── 33
│       └── pvc-05654189-ed2e-4995-a8e3-a7df775652e8
│           └── volume.cfg
├── 3f
│   └── 19
│       └── pvc-949549be-92b2-4105-b1ca-5e7d94ee6967
│           └── volume.cfg
├── 67
│   └── bd
│       └── pvc-c796c818-6b26-4985-bb8a-e4eecdeca80b
│           └── volume.cfg
└── 8b
    └── 7e
        └── pvc-7f1f8921-a2e9-4b96-bafa-12e080237bc2
            ├── backups
            │   └── backup_backup-1be66426a4a84768.cfg
            ├── blocks
            │   ├── 09
            │   │   └── 09
            │   │       └── 0909430cc9e0206f0ccef9cac6912db23f2795ed8460c4696802551883a5d6fd.blk
            │   ├── 0e
            │   │   └── b7
            │   │       └── 0eb7e9608b8bb5a6e24d81f4c82c76d6d2e13bfa2d2ee21af6905ec0ad19655a.blk
            │   ├── 2a
            │   │   └── 2c
            │   │       └── 2a2ca92b81083dfa363aa7cb637bf15ce5c502eb33aa071148430c9e616629b6.blk
            │   ├── 30
            │   │   └── 65
            │   │       └── 306543c5ebee58116b5c03a56d9d89a4f74fe1559f39b8cfe8d873dc36f9cd44.blk
            │   ├── 31
            │   │   └── 44
            │   │       └── 3144acd2a013b6c7c3ae10bbb2433711ee2c35e6cc7ba6e95ddb2a1ecb8236a2.blk
            │   ├── 32
            │   │   ├── dc
            │   │   │   └── 32dce3a9478be39113eeea9c29b9d7b966f5f6deb8889e4c6dd00bc7e0b25e88.blk
            │   │   └── f5
            │   │       └── 32f523e4123b193a6805ab38868ebc1a05028f02120549630843e47ea54f0821.blk
            │   ├── 49
            │   │   └── a8
            │   │       └── 49a8b26f188178c14ad3d7fb4373327f6013687c04cbcd8ea735ae551287a33c.blk
            │   ├── 4f
            │   │   └── 5c
            │   │       └── 4f5c2be091e32127900a87471c08acfa9b33872ad68d9f6fbf5287121c04ff44.blk
            │   ├── 6a
            │   │   └── c2
            │   │       └── 6ac2d8cf22aed855552ed46e9b275ebf012d3e0847f59b7c01603ecbaccb7407.blk
            │   ├── 6e
            │   │   └── db
            │   │       └── 6edbe2e7ec16a3392cbe5fb64811b41ef80208e8c24d5a9e94c119ecb7b8a609.blk
            │   ├── 6f
            │   │   └── 5f
            │   │       └── 6f5f302bc3724c926c359ea8a512b25c23dad18a998c656ef5ce3e0cf78769b6.blk
            │   ├── 73
            │   │   └── 18
            │   │       └── 731859029215873fdac1c9f2f8bd25a334abf0f3a9e1b057cf2cacc2826d86b0.blk
            │   ├── 7f
            │   │   └── 01
            │   │       └── 7f01e81990593fb74e04bfebf46290ced49f04d5a0666b1fdc7d2900349f5e45.blk
            │   ├── 84
            │   │   └── 67
            │   │       └── 8467beb0d28f6e1d80c0ae1435c76d4dc8f99ee629f375d0b17e52b870e08046.blk
            │   ├── 8b
            │   │   └── 01
            │   │       └── 8b010d5e48b0e00c65aebf3d93b990aa07bf9ba47113080873b557dbcded8179.blk
            │   ├── b0
            │   │   └── 00
            │   │       └── b0009ff1f5a8509cb515ad7772cde909313eb51123220aa7eaa3d045d50756cf.blk
            │   ├── bd
            │   │   └── 16
            │   │       └── bd163984b2403046d2a3fab25167dd4f6fdc7e30737d7e2a652b85825b4adc85.blk
            │   ├── c3
            │   │   └── b6
            │   │       └── c3b6bcf5b4a282384899462546690d761b97bcf43f93848cc0c5367111ce5656.blk
            │   ├── ca
            │   │   └── 0a
            │   │       └── ca0a79ee4319b904917617dd1200b675df4d6de99bb416b0c44e7ec25edb6083.blk
            │   ├── cb
            │   │   └── 87
            │   │       └── cb878b09ecaf92970d963a37b9546a0d987a04962d25e8c203f2e346c57e79ca.blk
            │   ├── d9
            │   │   └── 3b
            │   │       └── d93bcf462676092c7f35103dda78520c0aadad545f0097bf03ad397019167734.blk
            │   ├── e6
            │   │   └── c1
            │   │       └── e6c12c091c582ff84a57e85245d50a832256e3a869c8fc2810398049d88fa809.blk
            │   ├── f2
            │   │   └── 6b
            │   │       └── f26bcc2505a15db7a2012dcd51e8ceb3e49f5d9adf9a6fac63af0824efc02b76.blk
            │   ├── f3
            │   │   └── 0d
            │   │       └── f30d84ec1ce95998b68cfe336dbbc29ce5b9ed419283aa3cba8c5cf9b8096f85.blk
            │   └── f9
            │       ├── 7d
            │       │   └── f97d41cc22546b9508f14cbd173961c351f31b1c7f5a4b149918ad6465ed9f06.blk
            │       └── f8
            │           └── f9f830318b38066f5d17c406690bab6bfda7793e0b7170829bac75a75ef2b46a.blk
            └── volume.cfg

73 directories, 35 files
```
2. delete the "backup", after a while, check NFS server, most of the related files are removed


```
rancher@ubuntuvmday0:~$ sudo tree /nfs_server/nfs_share/backupstore/volumes --timefmt=%A
/nfs_server/nfs_share/backupstore/volumes
├── 17
│   └── 91
│       └── pvc-0a7cd9e3-b847-481b-a369-9f22686dee8f
│           └── volume.cfg
├── 2d
│   └── d2
│       └── pvc-4bc23108-d2cb-44b8-8617-6d11248b2593
│           └── volume.cfg
├── 36
│   └── 33
│       └── pvc-05654189-ed2e-4995-a8e3-a7df775652e8
│           └── volume.cfg
├── 3f
│   └── 19
│       └── pvc-949549be-92b2-4105-b1ca-5e7d94ee6967
│           └── volume.cfg
├── 67
│   └── bd
│       └── pvc-c796c818-6b26-4985-bb8a-e4eecdeca80b
│           └── volume.cfg
└── 8b
    └── 7e
        └── pvc-7f1f8921-a2e9-4b96-bafa-12e080237bc2
            └── volume.cfg
18 directories, 6 files
```

3. The remaining path and file, is `"backupstore/volumes/8b/7e/pvc-7f1f8921-a2e9-4b96-bafa-12e080237bc2/volume.cfg"`

As per current implementation, this is a meta-data of backup, is not deleted when backup is deleted.

rancher@ubuntuvmday0:~$ sudo cat /nfs_server/nfs_share/backupstore/volumes/8b/7e/pvc-7f1f8921-a2e9-4b96-bafa-12e080237bc2/volume.cfg

```
{"Name":"pvc-7f1f8921-a2e9-4b96-bafa-12e080237bc2","Size":"1073741824",

"Labels":{"KubernetesStatus":"{\"pvName\":\"pvc-7f1f8921-a2e9-4b96-bafa-12e080237bc2\",\"pvStatus\":\"Bound\",\"namespace\":\"default\",\"pvcName\":\"cirros-1801-test-disk-2-cd-myggi\",\"lastPVCRefAt\":\"\",\"workloadsStatus\":[{\"podName\":\"virt-launcher-cirros-1801-test-hqlfb\",\"podStatus\":\"Running\",\"workloadName\":\"cirros-1801-test\",\"workloadType\":\"VirtualMachineInstance\"}],\"lastPodRefAt\":\"\"}"},

"CreatedTime":"2022-01-18T14:12:17Z","LastBackupName":"","LastBackupAt":"","BlockCount":"0","BackingImageName":"\"default-image-mxcmj\"",
"BackingImageChecksum":"\"6b813aa46bb90b4da216a4d19376593fa3f4fc7e617f03a92b7fe11e9a3981cbe8f0959dbebe36225e5f53dc4492341a4863cac4ed1ee0909f3fc78ef9c3e869\""}

```

**Comment by w13915984028** [comment]
As per test in v1.0.0, after `deleting a backup instance`, path and file like `"backupstore/volumes/8b/7e/pvc-7f1f8921-a2e9-4b96-bafa-12e080237bc2/volume.cfg"` will be left,

@johnliu55tw 
Please also check, if only that path and file are left following your test steps. thanks.

@FrankYang0529  
Please confirm, if only the volume.cfg is left, what is the purpose of it, when will this file be finally deleted, or it is left forever, thanks.

**Comment by w13915984028** [comment]
Will check, why longhorn side leaves backup meta file like "backupstore/volumes/8b/7e/pvc-7f1f8921-a2e9-4b96-bafa-12e080237bc2/volume.cfg" in the backup server, and also has those "empty" backup instances
cc @joshimoo @PhanLe1010

`Harvester backup: empty`
![image](https://user-images.githubusercontent.com/31133476/149969326-6d3b25b6-cf84-4ffa-b239-ba47d91f2671.png)

`Longhorn backup: some instances, can be mapped to NFS server side remaining files`
![image](https://user-images.githubusercontent.com/31133476/149969081-a6957d94-476b-45b9-b86f-bbec109883cd.png)

`But the instance, has no useful information`
![image](https://user-images.githubusercontent.com/31133476/149970211-38813569-4571-4919-84b2-a260db1378d5.png)


`NFS backup target: remaining files`
![image](https://user-images.githubusercontent.com/31133476/149969210-fd3d5cb6-776b-4d89-aff3-657b22c0f24b.png)


`longhorn backup setting`
![image](https://user-images.githubusercontent.com/31133476/149970622-6a8e3a29-b49e-49f8-8113-75382e37dc31.png)

**Comment by PhanLe1010** [comment]
In Longhorn there are 2 objects: backup volume and backup. A backup volume is a collection of backups. When users delete the last backup of the volume, I don't think Longhorn automatically cleans up the backup volume. Therefore, you see the above behavior. 

What is the problem Harvester is having with behavior? The empty backup volume doesn't consume much data.

**Comment by w13915984028** [comment]
@PhanLe1010 
From the Harvester UI, the is no "backup" objects, all are deleted, then the user is a bit confused, why in the backup server, the is still something there.

The general expectation is, when a "backup" object is deleted, there should also be nothing related remaining in the backup server.

**Comment by PhanLe1010** [comment]
@w13915984028  

This will need a design discussion with the team. cc @joshimoo, @jenting 
Some ideas:
* Do we want Longhorn to auto-delete the backup volume if the last backup is deleted?
* Can Harvester delete the backup volume if the last backup is deleted?

**Comment by joshimoo** [comment]
We previously before the lock stuff, and all the other backup changes switched from "auto delete the backup volume" when the last backup is deleted due to: 
- concurrency issues
- backup deletion requires a GC pass while the backup volume deletion is just a flat out remove everything.

Since we already had the code and exposed the backup delete same as delete backup volume, we clarified the expectation that these are separate operations. Which makes sense since they are separate operations Independent from the backup format. (assume a backup format, that doesn't even have the volume.cfg for example)

From the harvester perspective if one wants to delete all the backups for a volume, then it's best to go via the backup volume delete approach since it does not require GC and neither eval loop for the deletion.

Check with @jenting but I think it's sufficient to delete the backupVolume crd at the moment, that should trigger the backup volume delete command.

**Comment by jenting** [comment]
> From the harvester perspective if one wants to delete all the backups for a volume, then it's best to go via the backup volume delete approach since it does not require GC and neither eval loop for the deletion.

Agree. From Harvester usage/perspective, if we want to delete all volume's backup, it's better to call backup volume deletion instead of looping all backups deletion because the performance is better (only one operation).
Or Harvester could auto delete the backup volume CRD once the last backup is deleted.

Since we consider [removing the backup volume CRD](https://github.com/longhorn/longhorn/issues/3375) in the future. After that, the last backup is deleted, the corresponding metadata will be deleted as well (probably drop the `volume.cfg` _or_ auto delete `volume.cfg` once the last backup is deleted).
